In [1]:
import pandas as pd
import numpy as np
import json

In [2]:
import warnings 
warnings.filterwarnings('ignore')

# ETL PROCESS START 
## 1) Extract

## Source 2 — Excel load karo

In [3]:
sales_marketing = pd.read_excel(
    r"D:\Data Science Learning\data_warehource project\source2,3,4\sales_marketing.xlsx"
)

print("Source 2 - Sales & Marketing")
print(sales_marketing.head())
print("Rows:", len(sales_marketing))

Source 2 - Sales & Marketing
                           order_id  Sales_Target  Discount_Percent  \
0  e481f51cbdc54678b7cc49136f2d6af7          1326                10   
1  53cdb2fc8bc7dce0b6741e2150273451          1659                10   
2  47770eb9100c2d0c44946d9cf07ec65d          1060                 5   
3  949d5b44dbf5de918fe9c16f97b45f8a          1494                15   
4  ad21c59c0840e6cb83a9ceb5573f8159          1330                 0   

  Marketing_Channel  Campaign_Cost Customer_Segment  
0             Email            285          Premium  
1          Facebook            273          Premium  
2            Google            252          Regular  
3         Instagram            305          Premium  
4             Email            135          Regular  
Rows: 1000


## Source 3 — CSV load karo

In [4]:
delivery_support = pd.read_csv(
    r"D:\Data Science Learning\data_warehource project\source2,3,4\delivery_support.csv"
)

print("\nSource 3 - Delivery & Support")
print(delivery_support.head())
print("Rows:", len(delivery_support))


Source 3 - Delivery & Support
                           order_id  Delivery_Days Shipping_Method  \
0  e481f51cbdc54678b7cc49136f2d6af7              9        Same Day   
1  53cdb2fc8bc7dce0b6741e2150273451              9        Same Day   
2  47770eb9100c2d0c44946d9cf07ec65d              4         Express   
3  949d5b44dbf5de918fe9c16f97b45f8a              8        Same Day   
4  ad21c59c0840e6cb83a9ceb5573f8159              8         Express   

   Support_Tickets  Customer_Satisfaction  
0                2                      2  
1                0                      5  
2                0                      1  
3                1                      2  
4                3                      2  
Rows: 1000


## Source 4 — JSON load karo

In [5]:
customer_risk = pd.read_json(
    r"D:\Data Science Learning\data_warehource project\source2,3,4\customer_risk.json"
)

print("\nSource 4 - Customer Risk")
print(customer_risk.head())
print("Rows:", len(customer_risk))


Source 4 - Customer Risk
                           Order_ID  Fraud_Risk_Score  Return_Flag  \
0  e481f51cbdc54678b7cc49136f2d6af7              0.95            0   
1  53cdb2fc8bc7dce0b6741e2150273451              0.23            0   
2  47770eb9100c2d0c44946d9cf07ec65d              0.59            0   
3  949d5b44dbf5de918fe9c16f97b45f8a              0.43            1   
4  ad21c59c0840e6cb83a9ceb5573f8159              0.76            0   

   Customer_Lifetime_Value Risk_Category  
0                     2342          High  
1                     8471           Low  
2                     7173        Medium  
3                     8913        Medium  
4                     7020          High  
Rows: 1000


## Quick check

In [6]:
print("\n--- Column Names ---")

print("Sales Marketing:")
print(sales_marketing.columns.tolist())

print("\nDelivery Support:")
print(delivery_support.columns.tolist())

print("\nCustomer Risk:")
print(customer_risk.columns.tolist())


--- Column Names ---
Sales Marketing:
['order_id', 'Sales_Target', 'Discount_Percent', 'Marketing_Channel', 'Campaign_Cost', 'Customer_Segment']

Delivery Support:
['order_id', 'Delivery_Days', 'Shipping_Method', 'Support_Tickets', 'Customer_Satisfaction']

Customer Risk:
['Order_ID', 'Fraud_Risk_Score', 'Return_Flag', 'Customer_Lifetime_Value', 'Risk_Category']


## now we load data from postgreSQL 

In [7]:
!pip install psycopg2-binary

Defaulting to user installation because normal site-packages is not writeable


In [8]:
import psycopg2

- connect PostgreSQL 

In [9]:
conn = psycopg2.connect(
    host="localhost",
    port="5432",
    database="shopsphere_source",
    user="postgres",
    password="Naveed2129"
)

print("PostgreSQL connection successful!")

PostgreSQL connection successful!


- PostgreSQL tables Python mein load karnam


In [10]:
customers = pd.read_sql(
    "SELECT * FROM olist_customers_dataset",
    conn
)

products = pd.read_sql(
    "SELECT * FROM olist_products_dataset",
    conn
)

sellers = pd.read_sql(
    "SELECT * FROM olist_sellers_dataset",
    conn
)

orders = pd.read_sql(
    "SELECT * FROM olist_orders_dataset",
    conn
)

order_items = pd.read_sql(
    "SELECT * FROM olist_order_items_dataset",
    conn
)

payments = pd.read_sql(
    "SELECT * FROM olist_order_payments_dataset",
    conn
)

print("Source 1 successfully extracted!")

Source 1 successfully extracted!


- now check the row count 

In [11]:
print("Customers:", len(customers))
print("Products:", len(products))
print("Sellers:", len(sellers))
print("Orders:", len(orders))
print("Order Items:", len(order_items))
print("Payments:", len(payments))

Customers: 99441
Products: 32951
Sellers: 3095
Orders: 99441
Order Items: 112650
Payments: 103886


In [12]:
print("\nCustomers:")
print(customers.head())

print("\nProducts:")
print(products.head())

print("\nOrders:")
print(orders.head())


Customers:
                        customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2  4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3  b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4  4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   

   customer_zip_code_prefix          customer_city customer_state  
0                     14409                 franca             SP  
1                      9790  sao bernardo do campo             SP  
2                      1151              sao paulo             SP  
3                      8775        mogi das cruzes             SP  
4                     13056               campinas             SP  

Products:
                         product_id  product_category_name  \
0  1e9e8ef04dbcff4541ed26657ea517e5             perfumaria   
1  3aa

_____________
## Transform

- Step 1 — check columns of all sources 

In [13]:
print("=== Source 1 ===")
print("Customers:", customers.columns.tolist())
print("Products:", products.columns.tolist())
print("Sellers:", sellers.columns.tolist())
print("Orders:", orders.columns.tolist())
print("Order Items:", order_items.columns.tolist())
print("Payments:", payments.columns.tolist())

print("\n=== Source 2 ===")
print(sales_marketing.columns.tolist())

print("\n=== Source 3 ===")
print(delivery_support.columns.tolist())

print("\n=== Source 4 ===")
print(customer_risk.columns.tolist())

=== Source 1 ===
Customers: ['customer_id', 'customer_unique_id', 'customer_zip_code_prefix', 'customer_city', 'customer_state']
Products: ['product_id', 'product_category_name', 'product_name_lenght', 'product_description_lenght', 'product_photos_qty', 'product_weight_g', 'product_length_cm', 'product_height_cm', 'product_width_cm']
Sellers: ['seller_id', 'seller_zip_code_prefix', 'seller_city', 'seller_state']
Orders: ['order_id', 'customer_id', 'order_status', 'order_purchase_timestamp', 'order_approved_at', 'order_delivered_carrier_date', 'order_delivered_customer_date', 'order_estimated_delivery_date']
Order Items: ['order_id', 'order_item_id', 'product_id', 'seller_id', 'shipping_limit_date', 'price', 'freight_value']
Payments: ['order_id', 'payment_sequential', 'payment_type', 'payment_installments', 'payment_value']

=== Source 2 ===
['order_id', 'Sales_Target', 'Discount_Percent', 'Marketing_Channel', 'Campaign_Cost', 'Customer_Segment']

=== Source 3 ===
['order_id', 'Deliver

- Step 2 — Source 1 Transformation

In [14]:
# TRANSFORM - SOURCE 1
# ==========================================

# Customers
customers = customers.rename(columns={
    "customer_zip_code_prefix": "zip_code",
    "customer_city": "city",
    "customer_state": "state"
})

# Products
products = products.rename(columns={
    "product_category_name": "category",
    "product_weight_g": "weight",
    "product_length_cm": "length",
    "product_height_cm": "height",
    "product_width_cm": "width"
})

# Sellers
sellers = sellers.rename(columns={
    "seller_zip_code_prefix": "zip_code",
    "seller_city": "city",
    "seller_state": "state"
})

# Orders
orders["order_purchase_timestamp"] = pd.to_datetime(
    orders["order_purchase_timestamp"],
    errors="coerce"
)

# Payments
payments = payments.rename(columns={
    "payment_installments": "installments"
})

print("Source 1 transformation completed!")

print("\nCustomers:")
print(customers.head())

print("\nProducts:")
print(products.head())

print("\nSellers:")
print(sellers.head())

print("\nOrders:")
print(orders[["order_id", "customer_id", "order_purchase_timestamp"]].head())

print("\nPayments:")
print(payments.head())

Source 1 transformation completed!

Customers:
                        customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2  4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3  b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4  4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   

   zip_code                   city state  
0     14409                 franca    SP  
1      9790  sao bernardo do campo    SP  
2      1151              sao paulo    SP  
3      8775        mogi das cruzes    SP  
4     13056               campinas    SP  

Products:
                         product_id               category  \
0  1e9e8ef04dbcff4541ed26657ea517e5             perfumaria   
1  3aa071139cb16b67ca9e5dea641aaa2f                  artes   
2  96bd76ec8810374ed1b65e291975717f          esporte_lazer 

- Step 3 — Source 2, 3, 4 Transformation

In [15]:
# ==========================================
# TRANSFORM - SOURCE 2, 3, 4
# ==========================================

# ------------------------------------------
# Source 2 - Sales & Marketing
# ------------------------------------------

sales_marketing = sales_marketing.rename(columns={
    "Sales_Target": "sales_target",
    "Discount_Percent": "discount_percent",
    "Marketing_Channel": "marketing_channel",
    "Campaign_Cost": "campaign_cost",
    "Customer_Segment": "customer_segment"
})

# ------------------------------------------
# Source 3 - Delivery & Support
# ------------------------------------------

delivery_support = delivery_support.rename(columns={
    "Delivery_Days": "delivery_days",
    "Shipping_Method": "shipping_method",
    "Support_Tickets": "support_tickets",
    "Customer_Satisfaction": "customer_satisfaction"
})

# ------------------------------------------
# Source 4 - Customer Risk
# ------------------------------------------

customer_risk = customer_risk.rename(columns={
    "Order_ID": "order_id",
    "Fraud_Risk_Score": "fraud_risk_score",
    "Return_Flag": "return_flag",
    "Customer_Lifetime_Value": "customer_lifetime_value",
    "Risk_Category": "risk_category"
})

# ------------------------------------------
# Convert numeric columns
# ------------------------------------------

sales_marketing["sales_target"] = pd.to_numeric(
    sales_marketing["sales_target"], errors="coerce"
)

sales_marketing["discount_percent"] = pd.to_numeric(
    sales_marketing["discount_percent"], errors="coerce"
)

sales_marketing["campaign_cost"] = pd.to_numeric(
    sales_marketing["campaign_cost"], errors="coerce"
)

delivery_support["delivery_days"] = pd.to_numeric(
    delivery_support["delivery_days"], errors="coerce"
)

delivery_support["support_tickets"] = pd.to_numeric(
    delivery_support["support_tickets"], errors="coerce"
)

delivery_support["customer_satisfaction"] = pd.to_numeric(
    delivery_support["customer_satisfaction"], errors="coerce"
)

customer_risk["fraud_risk_score"] = pd.to_numeric(
    customer_risk["fraud_risk_score"], errors="coerce"
)

customer_risk["return_flag"] = pd.to_numeric(
    customer_risk["return_flag"], errors="coerce"
)

customer_risk["customer_lifetime_value"] = pd.to_numeric(
    customer_risk["customer_lifetime_value"], errors="coerce"
)

print("Source 2, 3 and 4 transformation completed!")

print("\n=== Source 2 ===")
print(sales_marketing.head())

print("\n=== Source 3 ===")
print(delivery_support.head())

print("\n=== Source 4 ===")
print(customer_risk.head())

Source 2, 3 and 4 transformation completed!

=== Source 2 ===
                           order_id  sales_target  discount_percent  \
0  e481f51cbdc54678b7cc49136f2d6af7          1326                10   
1  53cdb2fc8bc7dce0b6741e2150273451          1659                10   
2  47770eb9100c2d0c44946d9cf07ec65d          1060                 5   
3  949d5b44dbf5de918fe9c16f97b45f8a          1494                15   
4  ad21c59c0840e6cb83a9ceb5573f8159          1330                 0   

  marketing_channel  campaign_cost customer_segment  
0             Email            285          Premium  
1          Facebook            273          Premium  
2            Google            252          Regular  
3         Instagram            305          Premium  
4             Email            135          Regular  

=== Source 3 ===
                           order_id  delivery_days shipping_method  \
0  e481f51cbdc54678b7cc49136f2d6af7              9        Same Day   
1  53cdb2fc8bc7dce0b6741e2150

- Step 4 — Data Quality Check

In [16]:
# ==========================================
# DATA QUALITY CHECK
# ==========================================

datasets = {
    "Customers": customers,
    "Products": products,
    "Sellers": sellers,
    "Orders": orders,
    "Order Items": order_items,
    "Payments": payments,
    "Sales & Marketing": sales_marketing,
    "Delivery & Support": delivery_support,
    "Customer Risk": customer_risk
}

for name, df in datasets.items():

    print("\n" + "=" * 60)
    print(name)
    print("=" * 60)

    print("Rows:", len(df))
    print("Columns:", len(df.columns))

    print("\nMissing Values:")
    print(df.isnull().sum())

    print("\nDuplicate Rows:", df.duplicated().sum())

    print("\nData Types:")
    print(df.dtypes)


Customers
Rows: 99441
Columns: 5

Missing Values:
customer_id           0
customer_unique_id    0
zip_code              0
city                  0
state                 0
dtype: int64

Duplicate Rows: 0

Data Types:
customer_id           object
customer_unique_id    object
zip_code               int64
city                  object
state                 object
dtype: object

Products
Rows: 32951
Columns: 9

Missing Values:
product_id                      0
category                      610
product_name_lenght           610
product_description_lenght    610
product_photos_qty            610
weight                          2
length                          2
height                          2
width                           2
dtype: int64

Duplicate Rows: 0

Data Types:
product_id                     object
category                       object
product_name_lenght           float64
product_description_lenght    float64
product_photos_qty            float64
weight                        floa

- Step 5 — Missing Values Handle Karna

In [17]:
# ==========================================
# HANDLE MISSING VALUES
# ==========================================

# Products
products["category"] = products["category"].fillna("Unknown")

products["weight"] = products["weight"].fillna(0)
products["length"] = products["length"].fillna(0)
products["height"] = products["height"].fillna(0)
products["width"] = products["width"].fillna(0)

# Orders
# Delivery dates are kept as NULL because
# missing delivery information can represent
# undelivered/cancelled orders.

print("Missing values handled!")

print("\nProducts missing values:")
print(products.isnull().sum())

print("\nOrders missing values:")
print(orders.isnull().sum())

Missing values handled!

Products missing values:
product_id                      0
category                        0
product_name_lenght           610
product_description_lenght    610
product_photos_qty            610
weight                          0
length                          0
height                          0
width                           0
dtype: int64

Orders missing values:
order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                 160
order_delivered_carrier_date     1783
order_delivered_customer_date    2965
order_estimated_delivery_date       0
dtype: int64


In [18]:
# ==========================================
# PREPARE DIM_CUSTOMER
# ==========================================

dim_customer_data = customers[
    [
        "customer_id",
        "city",
        "state",
        "zip_code"
    ]
].copy()

print("dim_customer data prepared!")

print("Rows:", len(dim_customer_data))
print("Columns:", dim_customer_data.columns.tolist())

print("\nFirst 5 rows:")
print(dim_customer_data.head())

dim_customer data prepared!
Rows: 99441
Columns: ['customer_id', 'city', 'state', 'zip_code']

First 5 rows:
                        customer_id                   city state  zip_code
0  06b8999e2fba1a1fbc88172c00ba8bc7                 franca    SP     14409
1  18955e83d337fd6b2def6b18a428ac77  sao bernardo do campo    SP      9790
2  4e7b3e00288586ebd08712fdd0374a03              sao paulo    SP      1151
3  b2b6027bc5c5109e529d4dc6358b12c3        mogi das cruzes    SP      8775
4  4f2d8ab171c80ec8364f7c12e35b23ad               campinas    SP     13056


- Step 6 — Prepare dim_product

In [19]:
# ==========================================
# PREPARE DIM_PRODUCT
# ==========================================

dim_product_data = products[
    [
        "product_id",
        "category",
        "weight",
        "length",
        "height",
        "width"
    ]
].copy()

print("dim_product data prepared!")

print("Rows:", len(dim_product_data))
print("Columns:", dim_product_data.columns.tolist())

print("\nFirst 5 rows:")
print(dim_product_data.head())

dim_product data prepared!
Rows: 32951
Columns: ['product_id', 'category', 'weight', 'length', 'height', 'width']

First 5 rows:
                         product_id               category  weight  length  \
0  1e9e8ef04dbcff4541ed26657ea517e5             perfumaria   225.0    16.0   
1  3aa071139cb16b67ca9e5dea641aaa2f                  artes  1000.0    30.0   
2  96bd76ec8810374ed1b65e291975717f          esporte_lazer   154.0    18.0   
3  cef67bcfe19066a932b7673e239eb23d                  bebes   371.0    26.0   
4  9dc1a7de274444849c219cff195d0b71  utilidades_domesticas   625.0    20.0   

   height  width  
0    10.0   14.0  
1    18.0   20.0  
2     9.0   15.0  
3     4.0   26.0  
4    17.0   13.0  


- Step 7 — Prepare dim_seller

In [20]:
# ==========================================
# PREPARE DIM_SELLER
# ==========================================

dim_seller_data = sellers[
    [
        "seller_id",
        "city",
        "state",
        "zip_code"
    ]
].copy()

print("dim_seller data prepared!")

print("Rows:", len(dim_seller_data))
print("Columns:", dim_seller_data.columns.tolist())

print("\nFirst 5 rows:")
print(dim_seller_data.head())

dim_seller data prepared!
Rows: 3095
Columns: ['seller_id', 'city', 'state', 'zip_code']

First 5 rows:
                          seller_id               city state  zip_code
0  3442f8959a84dea7ee197c632cb2df15           campinas    SP     13023
1  d1b65fc7debc3361ea86b5f14c68d2e2         mogi guacu    SP     13844
2  ce3ad9de960102d0677a81f5d0bb7b2d     rio de janeiro    RJ     20031
3  c0f3eea2e14555b6faeea3dd58c1b1c3          sao paulo    SP      4195
4  51a04a8a6bdcb23deccc82b0b80742cf  braganca paulista    SP     12914


- Step 8 — Prepare dim_payment

In [21]:
# ==========================================
# PREPARE DIM_PAYMENT
# ==========================================

dim_payment_data = payments[
    [
        "payment_type",
        "installments"
    ]
].copy()

# Remove exact duplicate combinations
dim_payment_data = dim_payment_data.drop_duplicates().reset_index(drop=True)

print("dim_payment data prepared!")

print("Rows:", len(dim_payment_data))
print("Columns:", dim_payment_data.columns.tolist())

print("\nFirst 10 rows:")
print(dim_payment_data.head(10))

dim_payment data prepared!
Rows: 28
Columns: ['payment_type', 'installments']

First 10 rows:
  payment_type  installments
0  credit_card             8
1  credit_card             1
2  credit_card             2
3  credit_card             3
4  credit_card             6
5       boleto             1
6  credit_card             5
7  credit_card             4
8  credit_card            10
9      voucher             1


- Step 9 — Prepare dim_date

In [22]:
# ==========================================
# PREPARE DIM_DATE
# ==========================================

# Purchase dates se unique dates nikalna
date_data = orders[
    ["order_purchase_timestamp"]
].copy()

date_data["full_date"] = date_data["order_purchase_timestamp"].dt.date

# Unique dates
date_data = date_data[["full_date"]].drop_duplicates().reset_index(drop=True)

# Date attributes
date_data["full_date"] = pd.to_datetime(date_data["full_date"])

date_data["date_key"] = date_data["full_date"].dt.strftime("%Y%m%d").astype(int)

date_data["day"] = date_data["full_date"].dt.day

date_data["month"] = date_data["full_date"].dt.month

date_data["month_name"] = date_data["full_date"].dt.month_name()

date_data["quarter"] = date_data["full_date"].dt.quarter

date_data["year"] = date_data["full_date"].dt.year

# Final column order
dim_date_data = date_data[
    [
        "date_key",
        "full_date",
        "day",
        "month",
        "month_name",
        "quarter",
        "year"
    ]
].copy()

print("dim_date data prepared!")

print("Rows:", len(dim_date_data))
print("Columns:", dim_date_data.columns.tolist())

print("\nFirst 10 rows:")
print(dim_date_data.head(10))

dim_date data prepared!
Rows: 634
Columns: ['date_key', 'full_date', 'day', 'month', 'month_name', 'quarter', 'year']

First 10 rows:
   date_key  full_date  day  month month_name  quarter  year
0  20171002 2017-10-02    2     10    October        4  2017
1  20180724 2018-07-24   24      7       July        3  2018
2  20180808 2018-08-08    8      8     August        3  2018
3  20171118 2017-11-18   18     11   November        4  2017
4  20180213 2018-02-13   13      2   February        1  2018
5  20170709 2017-07-09    9      7       July        3  2017
6  20170411 2017-04-11   11      4      April        2  2017
7  20170516 2017-05-16   16      5        May        2  2017
8  20170123 2017-01-23   23      1    January        1  2017
9  20170729 2017-07-29   29      7       July        3  2017


- Step 10 — Prepare dim_marketing

In [23]:
# ==========================================
# PREPARE DIM_MARKETING
# ==========================================

dim_marketing_data = sales_marketing[
    [
        "sales_target",
        "discount_percent",
        "marketing_channel",
        "campaign_cost",
        "customer_segment"
    ]
].copy()

# Remove exact duplicate combinations
dim_marketing_data = (
    dim_marketing_data
    .drop_duplicates()
    .reset_index(drop=True)
)

print("dim_marketing data prepared!")

print("Rows:", len(dim_marketing_data))
print("Columns:", dim_marketing_data.columns.tolist())

print("\nFirst 10 rows:")
print(dim_marketing_data.head(10))

dim_marketing data prepared!
Rows: 1000
Columns: ['sales_target', 'discount_percent', 'marketing_channel', 'campaign_cost', 'customer_segment']

First 10 rows:
   sales_target  discount_percent marketing_channel  campaign_cost  \
0          1326                10             Email            285   
1          1659                10          Facebook            273   
2          1060                 5            Google            252   
3          1494                15         Instagram            305   
4          1330                 0             Email            135   
5          1295                 0            Google            199   
6          1924                15          Facebook            284   
7          1244                 5            Google            394   
8          1838                10          Facebook             71   
9           321                 0         Instagram            397   

  customer_segment  
0          Premium  
1          Premium  
2     

- Step 11 — Prepare dim_delivery

In [24]:
# ==========================================
# PREPARE DIM_DELIVERY
# ==========================================

dim_delivery_data = delivery_support[
    [
        "delivery_days",
        "shipping_method",
        "support_tickets",
        "customer_satisfaction"
    ]
].copy()

# Remove exact duplicate combinations
dim_delivery_data = (
    dim_delivery_data
    .drop_duplicates()
    .reset_index(drop=True)
)

print("dim_delivery data prepared!")

print("Rows:", len(dim_delivery_data))
print("Columns:", dim_delivery_data.columns.tolist())

print("\nFirst 10 rows:")
print(dim_delivery_data.head(10))

dim_delivery data prepared!
Rows: 454
Columns: ['delivery_days', 'shipping_method', 'support_tickets', 'customer_satisfaction']

First 10 rows:
   delivery_days shipping_method  support_tickets  customer_satisfaction
0              9        Same Day                2                      2
1              9        Same Day                0                      5
2              4         Express                0                      1
3              8        Same Day                1                      2
4              8         Express                3                      2
5              1         Express                1                      5
6              5        Same Day                1                      3
7              3        Standard                3                      2
8              6        Same Day                2                      1
9              3         Express                0                      4


- Step 12 — Prepare dim_risk

In [25]:
# ==========================================
# PREPARE DIM_RISK
# ==========================================

dim_risk_data = customer_risk[
    [
        "fraud_risk_score",
        "return_flag",
        "customer_lifetime_value",
        "risk_category"
    ]
].copy()

# Remove exact duplicate combinations
dim_risk_data = (
    dim_risk_data
    .drop_duplicates()
    .reset_index(drop=True)
)

print("dim_risk data prepared!")

print("Rows:", len(dim_risk_data))
print("Columns:", dim_risk_data.columns.tolist())

print("\nFirst 10 rows:")
print(dim_risk_data.head(10))

dim_risk data prepared!
Rows: 1000
Columns: ['fraud_risk_score', 'return_flag', 'customer_lifetime_value', 'risk_category']

First 10 rows:
   fraud_risk_score  return_flag  customer_lifetime_value risk_category
0              0.95            0                     2342          High
1              0.23            0                     8471           Low
2              0.59            0                     7173        Medium
3              0.43            1                     8913        Medium
4              0.76            0                     7020          High
5              0.00            0                     5363           Low
6              0.36            0                     3807        Medium
7              0.91            1                     7268          High
8              0.46            0                     4192        Medium
9              0.98            0                     8990          High


### Fact Sales preparation

- Step 13 — Source 1 ko Join karke Fact Base Banana

first we  orders + customers + order_items + products + sellers + payments ko combine karenge

In [27]:
# ==========================================
# PREPARE FACT SALES - SOURCE 1
# ==========================================

fact_sales_data = order_items.copy()

# ------------------------------------------
# Join Orders
# ------------------------------------------

fact_sales_data = fact_sales_data.merge(
    orders[
        [
            "order_id",
            "customer_id",
            "order_purchase_timestamp"
        ]
    ],
    on="order_id",
    how="left"
)

# ------------------------------------------
# Join Customers
# ------------------------------------------

fact_sales_data = fact_sales_data.merge(
    customers[
        [
            "customer_id"
        ]
    ],
    on="customer_id",
    how="left"
)

# ------------------------------------------
# Join Products
# ------------------------------------------

fact_sales_data = fact_sales_data.merge(
    products[
        [
            "product_id"
        ]
    ],
    on="product_id",
    how="left"
)

# ------------------------------------------
# Join Sellers
# ------------------------------------------

fact_sales_data = fact_sales_data.merge(
    sellers[
        [
            "seller_id"
        ]
    ],
    on="seller_id",
    how="left"
)

# ------------------------------------------
# Join Payments
# ------------------------------------------

fact_sales_data = fact_sales_data.merge(
    payments[
        [
            "order_id",
            "payment_type",
            "installments"
        ]
    ],
    on="order_id",
    how="left"
)

print("Source 1 joins completed!")

print("\nRows:", len(fact_sales_data))
print("Columns:", len(fact_sales_data.columns))

print("\nColumns:")
print(fact_sales_data.columns.tolist())

print("\nFirst 5 rows:")
print(fact_sales_data.head())

Source 1 joins completed!

Rows: 117604
Columns: 11

Columns:
['order_id', 'order_item_id', 'product_id', 'seller_id', 'shipping_limit_date', 'price', 'freight_value', 'customer_id', 'order_purchase_timestamp', 'payment_type', 'installments']

First 5 rows:
                           order_id  order_item_id  \
0  00010242fe8c5a6d1ba2dd792cb16214              1   
1  00018f77f2f0320c557190d7a144bdd3              1   
2  000229ec398224ef6ca0657da4fc703e              1   
3  00024acbcdf0a6daa1e931b038114c75              1   
4  00042b26cf59d7ce69dfabb4e55b4fd9              1   

                         product_id                         seller_id  \
0  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1  e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
2  c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242adc84c38acab88f23d   
3  7634da152a4610f1595efa32f14722fc  9d7a1d34a5052409006425275ba1c2b4   
4  ac6c3623068f30de03045865e4e10089  df560393f3a51e7

aur yahan ek important ETL issue mil gaya, jo actually good hai because abhi warehouse mein load karne se pehle fix kar sakte hain.

order_items ke 112,650 rows the, lekin join ke baad 117,604 ho gaye. Reason: ek order_id ke against multiple payment records ho sakte hain. Direct payment join ne kuch order-item rows ko duplicate kar diya.

- Step 14 — Payment duplication fix

Hum payments ko pehle one row per order mein aggregate karenge, phir fact table se join karenge.

In [28]:
# ==========================================
# FIX PAYMENT DUPLICATION
# ==========================================

# One payment record per order
payment_summary = (
    payments
    .groupby("order_id", as_index=False)
    .agg(
        payment_type=("payment_type", "first"),
        installments=("installments", "max")
    )
)

print("Payment summary created!")
print("Payment rows:", len(payment_summary))

# ------------------------------------------
# Rebuild Fact Sales
# ------------------------------------------

fact_sales_data = order_items.copy()

# Join Orders
fact_sales_data = fact_sales_data.merge(
    orders[
        [
            "order_id",
            "customer_id",
            "order_purchase_timestamp"
        ]
    ],
    on="order_id",
    how="left"
)

# Join Payments - now one row per order
fact_sales_data = fact_sales_data.merge(
    payment_summary,
    on="order_id",
    how="left"
)

print("\nFact Sales rebuilt!")
print("Rows:", len(fact_sales_data))
print("Columns:", len(fact_sales_data.columns))

print("\nFirst 5 rows:")
print(fact_sales_data.head())

Payment summary created!
Payment rows: 99440

Fact Sales rebuilt!
Rows: 112650
Columns: 11

First 5 rows:
                           order_id  order_item_id  \
0  00010242fe8c5a6d1ba2dd792cb16214              1   
1  00018f77f2f0320c557190d7a144bdd3              1   
2  000229ec398224ef6ca0657da4fc703e              1   
3  00024acbcdf0a6daa1e931b038114c75              1   
4  00042b26cf59d7ce69dfabb4e55b4fd9              1   

                         product_id                         seller_id  \
0  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1  e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
2  c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242adc84c38acab88f23d   
3  7634da152a4610f1595efa32f14722fc  9d7a1d34a5052409006425275ba1c2b4   
4  ac6c3623068f30de03045865e4e10089  df560393f3a51e74553ab94004ba5c87   

  shipping_limit_date   price  freight_value  \
0 2017-09-19 09:45:35   58.90          13.29   
1 2017-05-03 11:05:13  239.90     

_______

Ab next important kaam hai Source 2, 3, 4 ko order_id ke through Fact Sales mein integrate karna. Isse hamari heterogeneous sources actually ek integrated warehouse dataset banengi.

- Step 15 — Join Source 2, 3, 4

In [29]:
# ==========================================
# INTEGRATE SOURCE 2, 3, 4
# ==========================================

# ------------------------------------------
# Join Source 2 - Sales & Marketing
# ------------------------------------------

fact_sales_data = fact_sales_data.merge(
    sales_marketing,
    on="order_id",
    how="left"
)

# ------------------------------------------
# Join Source 3 - Delivery & Support
# ------------------------------------------

fact_sales_data = fact_sales_data.merge(
    delivery_support,
    on="order_id",
    how="left"
)

# ------------------------------------------
# Join Source 4 - Customer Risk
# ------------------------------------------

fact_sales_data = fact_sales_data.merge(
    customer_risk,
    on="order_id",
    how="left"
)

print("Source 2, 3 and 4 integrated!")

print("\nRows:", len(fact_sales_data))
print("Columns:", len(fact_sales_data.columns))

print("\nColumns:")
print(fact_sales_data.columns.tolist())

print("\nFirst 5 rows:")
print(fact_sales_data.head())

Source 2, 3 and 4 integrated!

Rows: 112650
Columns: 24

Columns:
['order_id', 'order_item_id', 'product_id', 'seller_id', 'shipping_limit_date', 'price', 'freight_value', 'customer_id', 'order_purchase_timestamp', 'payment_type', 'installments', 'sales_target', 'discount_percent', 'marketing_channel', 'campaign_cost', 'customer_segment', 'delivery_days', 'shipping_method', 'support_tickets', 'customer_satisfaction', 'fraud_risk_score', 'return_flag', 'customer_lifetime_value', 'risk_category']

First 5 rows:
                           order_id  order_item_id  \
0  00010242fe8c5a6d1ba2dd792cb16214              1   
1  00018f77f2f0320c557190d7a144bdd3              1   
2  000229ec398224ef6ca0657da4fc703e              1   
3  00024acbcdf0a6daa1e931b038114c75              1   
4  00042b26cf59d7ce69dfabb4e55b4fd9              1   

                         product_id                         seller_id  \
0  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1  e5f2d52b802

output mein ek important cheez nazar aa rahi hai: first 5 rows mein Source 2/3/4 ki values NaN hain. Ye error zaroori nahi hai. Hamare generated sources mein sirf 1000 orders hain, jabke Olist fact mein 112,650 order-item rows hain. Pehle check karte hain ke kitne fact rows actually in 1000 orders se match hue.

- Step 16 — Integration Match Check

In [30]:
# ==========================================
# CHECK SOURCE 2, 3, 4 MATCHING
# ==========================================

print("=== Source 2 Match ===")
print(
    "Matched rows:",
    fact_sales_data["sales_target"].notna().sum()
)

print(
    "Unmatched rows:",
    fact_sales_data["sales_target"].isna().sum()
)

print("\n=== Source 3 Match ===")
print(
    "Matched rows:",
    fact_sales_data["delivery_days"].notna().sum()
)

print(
    "Unmatched rows:",
    fact_sales_data["delivery_days"].isna().sum()
)

print("\n=== Source 4 Match ===")
print(
    "Matched rows:",
    fact_sales_data["fraud_risk_score"].notna().sum()
)

print(
    "Unmatched rows:",
    fact_sales_data["fraud_risk_score"].isna().sum()
)

=== Source 2 Match ===
Matched rows: 1145
Unmatched rows: 111505

=== Source 3 Match ===
Matched rows: 1145
Unmatched rows: 111505

=== Source 4 Match ===
Matched rows: 1145
Unmatched rows: 111505


___________

Unique order_id = 1000
Fact table mein un orders ke multiple items hain
Isliye 1000 orders → 1,145 order-item rows match hue.

So:

Source	Matched fact rows	Unmatched
Source 2 — Marketing	1,145	111,505
Source 3 — Delivery	1,145	111,505
Source 4 — Risk	1,145	111,505

Important: Ye ETL error nahi hai. Humne left join kiya hai, isliye Olist ke baqi orders ke liye external source information NULL reh gayi.

Ab next step mein hum NULL ko warehouse-friendly values mein convert karenge. Lekin pehle ek important design decision: since these three sources only cover 1,000 orders, dimensions mein missing source data ko "Unknown" / 0 jaisi values se represent karna better hai rather than pretending that the information exists.

- Step 17 — Prepare External Source Values

In [31]:
# ==========================================
# HANDLE MISSING EXTERNAL SOURCE DATA
# ==========================================

# Source 2 - Marketing
fact_sales_data["sales_target"] = fact_sales_data["sales_target"].fillna(0)
fact_sales_data["discount_percent"] = fact_sales_data["discount_percent"].fillna(0)
fact_sales_data["marketing_channel"] = fact_sales_data["marketing_channel"].fillna("Unknown")
fact_sales_data["campaign_cost"] = fact_sales_data["campaign_cost"].fillna(0)
fact_sales_data["customer_segment"] = fact_sales_data["customer_segment"].fillna("Unknown")

# Source 3 - Delivery
fact_sales_data["delivery_days"] = fact_sales_data["delivery_days"].fillna(0)
fact_sales_data["shipping_method"] = fact_sales_data["shipping_method"].fillna("Unknown")
fact_sales_data["support_tickets"] = fact_sales_data["support_tickets"].fillna(0)
fact_sales_data["customer_satisfaction"] = fact_sales_data["customer_satisfaction"].fillna(0)

# Source 4 - Risk
fact_sales_data["fraud_risk_score"] = fact_sales_data["fraud_risk_score"].fillna(0)
fact_sales_data["return_flag"] = fact_sales_data["return_flag"].fillna(0)
fact_sales_data["customer_lifetime_value"] = fact_sales_data["customer_lifetime_value"].fillna(0)
fact_sales_data["risk_category"] = fact_sales_data["risk_category"].fillna("Unknown")

print("External source missing values handled!")

print("\nRemaining missing values:")
print(
    fact_sales_data[
        [
            "sales_target",
            "discount_percent",
            "marketing_channel",
            "delivery_days",
            "shipping_method",
            "fraud_risk_score",
            "risk_category"
        ]
    ].isnull().sum()
)

External source missing values handled!

Remaining missing values:
sales_target         0
discount_percent     0
marketing_channel    0
delivery_days        0
shipping_method      0
fraud_risk_score     0
risk_category        0
dtype: int64


Ab next step Fact table ke calculations karte hain.

Humein 3 important columns banani hain:

quantity → 1 because each row represents one product item
total_amount → price + freight_value
date_key → purchase date se YYYYMMDD

- Step 2 — Fact calculations

In [32]:
# ==========================================
# CREATE FACT TABLE CALCULATED COLUMNS
# ==========================================

# Each row represents one product item
fact_sales_data["quantity"] = 1

# Total amount for each item
fact_sales_data["total_amount"] = (
    fact_sales_data["price"] + fact_sales_data["freight_value"]
)

# Create date_key from order purchase date
fact_sales_data["date_key"] = (
    fact_sales_data["order_purchase_timestamp"]
    .dt.strftime("%Y%m%d")
    .astype(int)
)

print("Fact calculations completed!")

print("\nSelected columns:")
print(
    fact_sales_data[
        [
            "order_id",
            "product_id",
            "price",
            "freight_value",
            "quantity",
            "total_amount",
            "date_key"
        ]
    ].head()
)

Fact calculations completed!

Selected columns:
                           order_id                        product_id   price  \
0  00010242fe8c5a6d1ba2dd792cb16214  4244733e06e7ecb4970a6e2683c13e61   58.90   
1  00018f77f2f0320c557190d7a144bdd3  e5f2d52b802189ee658865ca93d83a8f  239.90   
2  000229ec398224ef6ca0657da4fc703e  c777355d18b72b67abbeef9df44fd0fd  199.00   
3  00024acbcdf0a6daa1e931b038114c75  7634da152a4610f1595efa32f14722fc   12.99   
4  00042b26cf59d7ce69dfabb4e55b4fd9  ac6c3623068f30de03045865e4e10089  199.90   

   freight_value  quantity  total_amount  date_key  
0          13.29         1         72.19  20170913  
1          19.93         1        259.83  20170426  
2          17.87         1        216.87  20180114  
3          12.79         1         25.78  20180808  
4          18.14         1        218.04  20170204  


Ab next step: Warehouse dimensions ko PostgreSQL mein load karna start karenge.

Lekin ek important cheez: pehle hum shopsphere_dw database se connection test karenge

______________
## LOAD

- Step 3 — Data Warehouse connection

In [33]:
# ==========================================
# CONNECT TO DATA WAREHOUSE
# ==========================================

import psycopg2

dw_conn = psycopg2.connect(
    host="localhost",
    port="5432",
    database="shopsphere_dw",
    user="postgres",
    password="Naveed2129"
)

print("Data Warehouse connection successful!")

Data Warehouse connection successful!


- Step 4 — dim_customer load 

In [34]:
# ==========================================
# LOAD CUSTOMER DIMENSION
# ==========================================

from psycopg2.extras import execute_values

customer_rows = list(
    dim_customer_data[
        ["customer_id", "city", "state", "zip_code"]
    ].itertuples(index=False, name=None)
)

customer_insert_query = """
    INSERT INTO dim_customer
    (customer_id, city, state, zip_code)
    VALUES %s
"""

with dw_conn.cursor() as cursor:
    execute_values(
        cursor,
        customer_insert_query,
        customer_rows
    )

dw_conn.commit()

print("dim_customer loaded successfully!")
print("Rows loaded:", len(customer_rows))

dim_customer loaded successfully!
Rows loaded: 99441


- Step 5 — dim_product load

In [36]:
# ==========================================
# LOAD PRODUCT DIMENSION
# ==========================================

product_rows = list(
    dim_product_data[
        ["product_id", "category", "weight", "length", "height", "width"]
    ].itertuples(index=False, name=None)
)

product_insert_query = """
    INSERT INTO dim_product
    (product_id, category, weight, length, height, width)
    VALUES %s
"""

with dw_conn.cursor() as cursor:
    execute_values(
        cursor,
        product_insert_query,
        product_rows
    )

dw_conn.commit()

print("dim_product loaded successfully!")
print("Rows loaded:", len(product_rows))

dim_product loaded successfully!
Rows loaded: 32951


- Step 6 — dim_seller load

In [37]:
# ==========================================
# LOAD SELLER DIMENSION
# ==========================================

seller_rows = list(
    dim_seller_data[
        ["seller_id", "city", "state", "zip_code"]
    ].itertuples(index=False, name=None)
)

seller_insert_query = """
    INSERT INTO dim_seller
    (seller_id, city, state, zip_code)
    VALUES %s
"""

with dw_conn.cursor() as cursor:
    execute_values(
        cursor,
        seller_insert_query,
        seller_rows
    )

dw_conn.commit()

print("dim_seller loaded successfully!")
print("Rows loaded:", len(seller_rows))

dim_seller loaded successfully!
Rows loaded: 3095


- Step 7 — dim_payment load

In [39]:
# ==========================================
# LOAD PAYMENT DIMENSION
# ==========================================

payment_rows = list(
    dim_payment_data[
        ["payment_type", "installments"]
    ].itertuples(index=False, name=None)
)

payment_insert_query = """
    INSERT INTO dim_payment
    (payment_type, installments)
    VALUES %s
"""

with dw_conn.cursor() as cursor:
    execute_values(
        cursor,
        payment_insert_query,
        payment_rows
    )

dw_conn.commit()

print("dim_payment loaded successfully!")
print("Rows loaded:", len(payment_rows))

dim_payment loaded successfully!
Rows loaded: 28


- Step 8 — dim_date

In [40]:
# ==========================================
# LOAD DATE DIMENSION
# ==========================================

date_rows = list(
    dim_date_data[
        [
            "date_key",
            "full_date",
            "day",
            "month",
            "month_name",
            "quarter",
            "year"
        ]
    ].itertuples(index=False, name=None)
)

date_insert_query = """
    INSERT INTO dim_date
    (date_key, full_date, day, month, month_name, quarter, year)
    VALUES %s
"""

with dw_conn.cursor() as cursor:
    execute_values(
        cursor,
        date_insert_query,
        date_rows
    )

dw_conn.commit()

print("dim_date loaded successfully!")
print("Rows loaded:", len(date_rows))

dim_date loaded successfully!
Rows loaded: 634


- Step 9 — dim_marketing load (soruce 2)

In [41]:
# ==========================================
# LOAD MARKETING DIMENSION
# ==========================================

marketing_rows = list(
    dim_marketing_data[
        [
            "sales_target",
            "discount_percent",
            "marketing_channel",
            "campaign_cost",
            "customer_segment"
        ]
    ].itertuples(index=False, name=None)
)

marketing_insert_query = """
    INSERT INTO dim_marketing
    (
        sales_target,
        discount_percent,
        marketing_channel,
        campaign_cost,
        customer_segment
    )
    VALUES %s
"""

with dw_conn.cursor() as cursor:
    execute_values(
        cursor,
        marketing_insert_query,
        marketing_rows
    )

dw_conn.commit()

print("dim_marketing loaded successfully!")
print("Rows loaded:", len(marketing_rows))

dim_marketing loaded successfully!
Rows loaded: 1000


- Step 10 — dim_delivery load ( source 3 csv)

In [42]:
# ==========================================
# LOAD DELIVERY DIMENSION
# ==========================================

delivery_rows = list(
    dim_delivery_data[
        [
            "delivery_days",
            "shipping_method",
            "support_tickets",
            "customer_satisfaction"
        ]
    ].itertuples(index=False, name=None)
)

delivery_insert_query = """
    INSERT INTO dim_delivery
    (
        delivery_days,
        shipping_method,
        support_tickets,
        customer_satisfaction
    )
    VALUES %s
"""

with dw_conn.cursor() as cursor:
    execute_values(
        cursor,
        delivery_insert_query,
        delivery_rows
    )

dw_conn.commit()

print("dim_delivery loaded successfully!")
print("Rows loaded:", len(delivery_rows))

dim_delivery loaded successfully!
Rows loaded: 454


- Step 11 — dim_risk load (soruce 4 json)

In [43]:
# ==========================================
# LOAD RISK DIMENSION
# ==========================================

risk_rows = list(
    dim_risk_data[
        [
            "fraud_risk_score",
            "return_flag",
            "customer_lifetime_value",
            "risk_category"
        ]
    ].itertuples(index=False, name=None)
)

risk_insert_query = """
    INSERT INTO dim_risk
    (
        fraud_risk_score,
        return_flag,
        customer_lifetime_value,
        risk_category
    )
    VALUES %s
"""

with dw_conn.cursor() as cursor:
    execute_values(
        cursor,
        risk_insert_query,
        risk_rows
    )

dw_conn.commit()

print("dim_risk loaded successfully!")
print("Rows loaded:", len(risk_rows))

dim_risk loaded successfully!
Rows loaded: 1000


__________
🎯 Ab saari 8 dimension tables load ho chuki hain.



Ab hum project ke sabse important part par aa rahe hain: fact_sales load karna.

Lekin fact mein directly IDs nahi dalenge. Humein pehle surrogate keys (customer_key, product_key, etc.) retrieve karke fact_sales_data mein map karna hoga.

- Step 12 — Pehle Customer Keys retrieve karo

In [44]:
# ==========================================
# GET CUSTOMER SURROGATE KEYS
# ==========================================

customer_keys = pd.read_sql(
    """
    SELECT customer_key, customer_id
    FROM dim_customer
    """,
    dw_conn
)

print("Customer keys retrieved!")
print(customer_keys.head())
print("Rows:", len(customer_keys))

Customer keys retrieved!
   customer_key                       customer_id
0             1  06b8999e2fba1a1fbc88172c00ba8bc7
1             2  18955e83d337fd6b2def6b18a428ac77
2             3  4e7b3e00288586ebd08712fdd0374a03
3             4  b2b6027bc5c5109e529d4dc6358b12c3
4             5  4f2d8ab171c80ec8364f7c12e35b23ad
Rows: 99441


Customer surrogate keys successfully retrieve ho gaye — 99,441 rows.

Ab hum in keys ko fact_sales_data ke andar map karenge.

- Step 13 — customer_key map karo

In [45]:
# ==========================================
# MAP CUSTOMER KEY TO FACT
# ==========================================

fact_sales_data = fact_sales_data.merge(
    customer_keys,
    on="customer_id",
    how="left"
)

print("Customer key mapped successfully!")

print(
    fact_sales_data[
        ["order_id", "customer_id", "customer_key"]
    ].head()
)

print("\nMissing customer keys:",
      fact_sales_data["customer_key"].isna().sum())

Customer key mapped successfully!
                           order_id                       customer_id  \
0  00010242fe8c5a6d1ba2dd792cb16214  3ce436f183e68e07877b285a838db11a   
1  00018f77f2f0320c557190d7a144bdd3  f6dd3ec061db4e3987629fe6b26e5cce   
2  000229ec398224ef6ca0657da4fc703e  6489ae5e4333f3693df5ad4372dab6d3   
3  00024acbcdf0a6daa1e931b038114c75  d4eb9395c8c0431ee92fce09860c5a06   
4  00042b26cf59d7ce69dfabb4e55b4fd9  58dbd0b2d70206bf40e62cd34e84d795   

   customer_key  
0         65558  
1         34266  
2         34956  
3         51764  
4          7603  

Missing customer keys: 0


 Ab next product surrogate key map karte hain.

- Step 14 — Product Key

In [46]:
# ==========================================
# GET PRODUCT SURROGATE KEYS
# ==========================================

product_keys = pd.read_sql(
    """
    SELECT product_key, product_id
    FROM dim_product
    """,
    dw_conn
)

print("Product keys retrieved!")
print(product_keys.head())
print("Rows:", len(product_keys))

Product keys retrieved!
   product_key                        product_id
0            1  1e9e8ef04dbcff4541ed26657ea517e5
1            2  3aa071139cb16b67ca9e5dea641aaa2f
2            3  96bd76ec8810374ed1b65e291975717f
3            4  cef67bcfe19066a932b7673e239eb23d
4            5  9dc1a7de274444849c219cff195d0b71
Rows: 32951


Product keys bhi 32,951 retrieve ho gayi hain.

Ab inko fact_sales_data mein map karte hain.

- Step 15 — product_key map karo

In [47]:
# ==========================================
# MAP PRODUCT KEY TO FACT
# ==========================================

fact_sales_data = fact_sales_data.merge(
    product_keys,
    on="product_id",
    how="left"
)

print("Product key mapped successfully!")

print(
    fact_sales_data[
        ["order_id", "product_id", "product_key"]
    ].head()
)

print("\nMissing product keys:",
      fact_sales_data["product_key"].isna().sum())

Product key mapped successfully!
                           order_id                        product_id  \
0  00010242fe8c5a6d1ba2dd792cb16214  4244733e06e7ecb4970a6e2683c13e61   
1  00018f77f2f0320c557190d7a144bdd3  e5f2d52b802189ee658865ca93d83a8f   
2  000229ec398224ef6ca0657da4fc703e  c777355d18b72b67abbeef9df44fd0fd   
3  00024acbcdf0a6daa1e931b038114c75  7634da152a4610f1595efa32f14722fc   
4  00042b26cf59d7ce69dfabb4e55b4fd9  ac6c3623068f30de03045865e4e10089   

   product_key  
0        25866  
1        27231  
2        22625  
3        15404  
4         8863  

Missing product keys: 0


- tep 16 — Seller keys retrieve karo

In [48]:
# ==========================================
# GET SELLER SURROGATE KEYS
# ==========================================

seller_keys = pd.read_sql(
    """
    SELECT seller_key, seller_id
    FROM dim_seller
    """,
    dw_conn
)

print("Seller keys retrieved!")
print(seller_keys.head())
print("Rows:", len(seller_keys))

Seller keys retrieved!
   seller_key                         seller_id
0           1  3442f8959a84dea7ee197c632cb2df15
1           2  d1b65fc7debc3361ea86b5f14c68d2e2
2           3  ce3ad9de960102d0677a81f5d0bb7b2d
3           4  c0f3eea2e14555b6faeea3dd58c1b1c3
4           5  51a04a8a6bdcb23deccc82b0b80742cf
Rows: 3095


Ab seller_key ko fact table mein map karte hain.

- Step 17 — seller_key map karo

In [49]:
# ==========================================
# MAP SELLER KEY TO FACT
# ==========================================

fact_sales_data = fact_sales_data.merge(
    seller_keys,
    on="seller_id",
    how="left"
)

print("Seller key mapped successfully!")

print(
    fact_sales_data[
        ["order_id", "seller_id", "seller_key"]
    ].head()
)

print("\nMissing seller keys:",
      fact_sales_data["seller_key"].isna().sum())

Seller key mapped successfully!
                           order_id                         seller_id  \
0  00010242fe8c5a6d1ba2dd792cb16214  48436dade18ac8b2bce089ec2a041202   
1  00018f77f2f0320c557190d7a144bdd3  dd7ddc04e1b6c2c614352b383efe2d36   
2  000229ec398224ef6ca0657da4fc703e  5b51032eddd242adc84c38acab88f23d   
3  00024acbcdf0a6daa1e931b038114c75  9d7a1d34a5052409006425275ba1c2b4   
4  00042b26cf59d7ce69dfabb4e55b4fd9  df560393f3a51e74553ab94004ba5c87   

   seller_key  
0         514  
1         472  
2        1825  
3        2024  
4        1598  

Missing seller keys: 0


Ab next date_key ko verify/map karte hain. date_key already fact_sales_data mein hai, aur dim_date mein bhi same key hai.

- Step 18 — Date keys verify karo

In [50]:
# ==========================================
# GET DATE KEYS FROM WAREHOUSE
# ==========================================

date_keys = pd.read_sql(
    """
    SELECT date_key
    FROM dim_date
    """,
    dw_conn
)

print("Date keys retrieved!")
print(date_keys.head())
print("Rows:", len(date_keys))

print(
    "\nFact rows with missing date_key:",
    fact_sales_data["date_key"].isna().sum()
)

print(
    "Fact rows with unmatched date_key:",
    (~fact_sales_data["date_key"].isin(date_keys["date_key"])).sum()
)

Date keys retrieved!
   date_key
0  20171002
1  20180724
2  20180808
3  20171118
4  20180213
Rows: 634

Fact rows with missing date_key: 0
Fact rows with unmatched date_key: 0


- Ab next payment_key map karna hai. Yahan ek important point hai: dim_payment mein key combination payment_type + installments hai, isliye dono columns par lookup karenge.

Step 19 — Payment keys retrieve karo

In [51]:
# ==========================================
# GET PAYMENT SURROGATE KEYS
# ==========================================

payment_keys = pd.read_sql(
    """
    SELECT payment_key, payment_type, installments
    FROM dim_payment
    """,
    dw_conn
)

print("Payment keys retrieved!")
print(payment_keys.head())
print("Rows:", len(payment_keys))

Payment keys retrieved!
   payment_key payment_type  installments
0            1  credit_card             8
1            2  credit_card             1
2            3  credit_card             2
3            4  credit_card             3
4            5  credit_card             6
Rows: 28


Ab inko fact_sales_data mein 2 columns ke basis par map karte hain:

payment_type
installments
- Step 20 — payment_key map karo

In [52]:
# ==========================================
# MAP PAYMENT KEY TO FACT
# ==========================================

fact_sales_data = fact_sales_data.merge(
    payment_keys,
    on=["payment_type", "installments"],
    how="left"
)

print("Payment key mapped successfully!")

print(
    fact_sales_data[
        [
            "order_id",
            "payment_type",
            "installments",
            "payment_key"
        ]
    ].head()
)

print(
    "\nMissing payment keys:",
    fact_sales_data["payment_key"].isna().sum()
)

Payment key mapped successfully!
                           order_id payment_type  installments  payment_key
0  00010242fe8c5a6d1ba2dd792cb16214  credit_card           2.0          3.0
1  00018f77f2f0320c557190d7a144bdd3  credit_card           3.0          4.0
2  000229ec398224ef6ca0657da4fc703e  credit_card           5.0          7.0
3  00024acbcdf0a6daa1e931b038114c75  credit_card           2.0          3.0
4  00042b26cf59d7ce69dfabb4e55b4fd9  credit_card           3.0          4.0

Missing payment keys: 409


Step 21 — Missing payment rows check karo

In [53]:
# ==========================================
# CHECK MISSING PAYMENT KEYS
# ==========================================

missing_payment = fact_sales_data[
    fact_sales_data["payment_key"].isna()
][
    [
        "order_id",
        "payment_type",
        "installments"
    ]
]

print("Rows with missing payment_key:",
      len(missing_payment))

print("\nMissing payment details:")
print(missing_payment.head(20))

print("\nNull payment_type:",
      missing_payment["payment_type"].isna().sum())

print("Null installments:",
      missing_payment["installments"].isna().sum())

Rows with missing payment_key: 409

Missing payment details:
                              order_id payment_type  installments
38    0016dfedd97fc2950e388d2971d718c7      voucher           5.0
1263  02ec4da9d03014f06d711d60eb37cc22      voucher           2.0
1713  03ecec245220b63fd7f68c1737ba99ba      voucher           2.0
1714  03ecec245220b63fd7f68c1737ba99ba      voucher           2.0
1827  0420dbc50fc554e303e4b2d6b39063f6      voucher           4.0
1828  0420dbc50fc554e303e4b2d6b39063f6      voucher           4.0
1829  0420dbc50fc554e303e4b2d6b39063f6      voucher           4.0
3558  0814daa4d12a646aeb73c429d5852f4d      voucher           2.0
3862  08bdbdeaf1626720051193300f93806e      voucher           3.0
3863  08bdbdeaf1626720051193300f93806e      voucher           3.0
3864  08bdbdeaf1626720051193300f93806e      voucher           3.0
3865  08bdbdeaf1626720051193300f93806e      voucher           3.0
4358  09d048d5bfb9dd1a82f696ce9863e9b0      voucher           6.0
4359  09d048d5b

In [56]:
# ==========================================
# RESET FAILED TRANSACTION
# ==========================================

dw_conn.rollback()

print("Transaction reset successfully!")

Transaction reset successfully!


- Step 21C — dim_payment rebuild

In [57]:
# ==========================================
# REBUILD PAYMENT DIMENSION
# ==========================================

with dw_conn.cursor() as cursor:
    cursor.execute("DELETE FROM dim_payment")

dw_conn.commit()

# Use the same payment summary that was used for the fact table
dim_payment_final = payment_summary[
    ["payment_type", "installments"]
].drop_duplicates().copy()

# Add one Unknown record for missing payment information
unknown_payment = pd.DataFrame({
    "payment_type": ["Unknown"],
    "installments": [0]
})

dim_payment_final = pd.concat(
    [dim_payment_final, unknown_payment],
    ignore_index=True
)

# Prepare rows
payment_rows = list(
    dim_payment_final[
        ["payment_type", "installments"]
    ].itertuples(index=False, name=None)
)

# Insert
payment_insert_query = """
    INSERT INTO dim_payment
    (payment_type, installments)
    VALUES %s
"""

with dw_conn.cursor() as cursor:
    execute_values(
        cursor,
        payment_insert_query,
        payment_rows
    )

dw_conn.commit()

print("dim_payment rebuilt successfully!")
print("Rows loaded:", len(payment_rows))

dim_payment rebuilt successfully!
Rows loaded: 41


- step 22 — Payment key dobara map karo

In [58]:
# ==========================================
# REMAP PAYMENT KEY
# ==========================================

# Remove old payment_key
fact_sales_data = fact_sales_data.drop(
    columns=["payment_key"]
)

# Get new payment keys
payment_keys = pd.read_sql(
    """
    SELECT payment_key, payment_type, installments
    FROM dim_payment
    """,
    dw_conn
)

# Replace missing payment values with Unknown
fact_sales_data["payment_type"] = (
    fact_sales_data["payment_type"].fillna("Unknown")
)

fact_sales_data["installments"] = (
    fact_sales_data["installments"].fillna(0)
)

# Make sure both sides use the same numeric type
fact_sales_data["installments"] = (
    fact_sales_data["installments"].astype(int)
)

payment_keys["installments"] = (
    payment_keys["installments"].astype(int)
)

# Map payment_key
fact_sales_data = fact_sales_data.merge(
    payment_keys,
    on=["payment_type", "installments"],
    how="left"
)

print("Payment key remapped successfully!")

print(
    fact_sales_data[
        [
            "order_id",
            "payment_type",
            "installments",
            "payment_key"
        ]
    ].head()
)

print(
    "\nMissing payment keys:",
    fact_sales_data["payment_key"].isna().sum()
)

Payment key remapped successfully!
                           order_id payment_type  installments  payment_key
0  00010242fe8c5a6d1ba2dd792cb16214  credit_card             2           29
1  00018f77f2f0320c557190d7a144bdd3  credit_card             3           30
2  000229ec398224ef6ca0657da4fc703e  credit_card             5           31
3  00024acbcdf0a6daa1e931b038114c75  credit_card             2           29
4  00042b26cf59d7ce69dfabb4e55b4fd9  credit_card             3           30

Missing payment keys: 0


- Step 23 — Marketing key retrieve + map

In [59]:
# ==========================================
# GET MARKETING SURROGATE KEYS
# ==========================================

marketing_keys = pd.read_sql(
    """
    SELECT
        marketing_key,
        sales_target,
        discount_percent,
        marketing_channel,
        campaign_cost,
        customer_segment
    FROM dim_marketing
    """,
    dw_conn
)

print("Marketing keys retrieved!")
print(marketing_keys.head())
print("Rows:", len(marketing_keys))

Marketing keys retrieved!
   marketing_key  sales_target  discount_percent marketing_channel  \
0              1        1326.0              10.0             Email   
1              2        1659.0              10.0          Facebook   
2              3        1060.0               5.0            Google   
3              4        1494.0              15.0         Instagram   
4              5        1330.0               0.0             Email   

   campaign_cost customer_segment  
0          285.0          Premium  
1          273.0          Premium  
2          252.0          Regular  
3          305.0          Premium  
4          135.0          Regular  
Rows: 1000


Ab marketing ke 5 attributes ke basis par marketing_key map karte hain.

- Step 24 — marketing_key map karo

In [60]:
# ==========================================
# MAP MARKETING KEY TO FACT
# ==========================================

fact_sales_data = fact_sales_data.merge(
    marketing_keys,
    on=[
        "sales_target",
        "discount_percent",
        "marketing_channel",
        "campaign_cost",
        "customer_segment"
    ],
    how="left"
)

print("Marketing key mapped successfully!")

print(
    fact_sales_data[
        [
            "order_id",
            "marketing_channel",
            "customer_segment",
            "marketing_key"
        ]
    ].head()
)

print(
    "\nMissing marketing keys:",
    fact_sales_data["marketing_key"].isna().sum()
)

Marketing key mapped successfully!
                           order_id marketing_channel customer_segment  \
0  00010242fe8c5a6d1ba2dd792cb16214           Unknown          Unknown   
1  00018f77f2f0320c557190d7a144bdd3           Unknown          Unknown   
2  000229ec398224ef6ca0657da4fc703e           Unknown          Unknown   
3  00024acbcdf0a6daa1e931b038114c75           Unknown          Unknown   
4  00042b26cf59d7ce69dfabb4e55b4fd9           Unknown          Unknown   

   marketing_key  
0            NaN  
1            NaN  
2            NaN  
3            NaN  
4            NaN  

Missing marketing keys: 111505


- Step 24A — dim_marketing mein Unknown record add karo

In [61]:
# ==========================================
# ADD UNKNOWN MARKETING RECORD
# ==========================================

with dw_conn.cursor() as cursor:
    cursor.execute("""
        INSERT INTO dim_marketing
        (
            sales_target,
            discount_percent,
            marketing_channel,
            campaign_cost,
            customer_segment
        )
        VALUES (0, 0, 'Unknown', 0, 'Unknown')
    """)

dw_conn.commit()

print("Unknown marketing record added!")

Unknown marketing record added!


- Step 24B — Marketing keys refresh karo

In [62]:
# ==========================================
# REFRESH MARKETING KEYS
# ==========================================

marketing_keys = pd.read_sql(
    """
    SELECT
        marketing_key,
        sales_target,
        discount_percent,
        marketing_channel,
        campaign_cost,
        customer_segment
    FROM dim_marketing
    """,
    dw_conn
)

print("Marketing keys refreshed!")
print("Rows:", len(marketing_keys))

print("\nUnknown record:")
print(
    marketing_keys[
        marketing_keys["marketing_channel"] == "Unknown"
    ]
)

Marketing keys refreshed!
Rows: 1001

Unknown record:
      marketing_key  sales_target  discount_percent marketing_channel  \
1000           1001           0.0               0.0           Unknown   

      campaign_cost customer_segment  
1000            0.0          Unknown  


- Step 24C — Marketing key remap

In [63]:
# ==========================================
# REMAP MARKETING KEY
# ==========================================

# Remove old marketing_key
fact_sales_data = fact_sales_data.drop(
    columns=["marketing_key"]
)

# Map marketing_key again
fact_sales_data = fact_sales_data.merge(
    marketing_keys,
    on=[
        "sales_target",
        "discount_percent",
        "marketing_channel",
        "campaign_cost",
        "customer_segment"
    ],
    how="left"
)

print("Marketing key remapped successfully!")

print(
    fact_sales_data[
        [
            "order_id",
            "marketing_channel",
            "customer_segment",
            "marketing_key"
        ]
    ].head()
)

print(
    "\nMissing marketing keys:",
    fact_sales_data["marketing_key"].isna().sum()
)

Marketing key remapped successfully!
                           order_id marketing_channel customer_segment  \
0  00010242fe8c5a6d1ba2dd792cb16214           Unknown          Unknown   
1  00018f77f2f0320c557190d7a144bdd3           Unknown          Unknown   
2  000229ec398224ef6ca0657da4fc703e           Unknown          Unknown   
3  00024acbcdf0a6daa1e931b038114c75           Unknown          Unknown   
4  00042b26cf59d7ce69dfabb4e55b4fd9           Unknown          Unknown   

   marketing_key  
0           1001  
1           1001  
2           1001  
3           1001  
4           1001  

Missing marketing keys: 0


Next step: Delivery Key

Pehle dim_delivery mein Unknown record add karte hain, kyunki fact mein unmatched orders ke liye delivery_days=0 aur shipping_method='Unknown' etc. hain.

In [64]:
# ==========================================
# ADD UNKNOWN DELIVERY RECORD
# ==========================================

with dw_conn.cursor() as cursor:
    cursor.execute("""
        INSERT INTO dim_delivery
        (
            delivery_days,
            shipping_method,
            support_tickets,
            customer_satisfaction
        )
        VALUES (0, 'Unknown', 0, 0)
    """)

dw_conn.commit()

print("Unknown delivery record added successfully!")

Unknown delivery record added successfully!


In [65]:
# ==========================================
# RETRIEVE DELIVERY KEYS
# ==========================================

delivery_keys = pd.read_sql(
    """
    SELECT
        delivery_key,
        delivery_days,
        shipping_method,
        support_tickets,
        customer_satisfaction
    FROM dim_delivery
    """,
    dw_conn
)

print("Delivery dimension rows:", len(delivery_keys))
print(delivery_keys.tail())

Delivery dimension rows: 455
     delivery_key  delivery_days shipping_method  support_tickets  \
450           451              5         Express                0   
451           452              9         Express                0   
452           453              7         Express                2   
453           454              2        Same Day                1   
454           455              0         Unknown                0   

     customer_satisfaction  
450                      2  
451                      1  
452                      3  
453                      1  
454                      0  


In [66]:
# ==========================================
# REMAP DELIVERY KEY
# ==========================================

fact_sales_data = fact_sales_data.drop(
    columns=["delivery_key"],
    errors="ignore"
)

fact_sales_data = fact_sales_data.merge(
    delivery_keys,
    on=[
        "delivery_days",
        "shipping_method",
        "support_tickets",
        "customer_satisfaction"
    ],
    how="left"
)

print("Delivery key mapped successfully!")

print(
    fact_sales_data[
        [
            "order_id",
            "delivery_days",
            "shipping_method",
            "delivery_key"
        ]
    ].head()
)

print(
    "\nMissing delivery keys:",
    fact_sales_data["delivery_key"].isna().sum()
)

Delivery key mapped successfully!
                           order_id  delivery_days shipping_method  \
0  00010242fe8c5a6d1ba2dd792cb16214            0.0         Unknown   
1  00018f77f2f0320c557190d7a144bdd3            0.0         Unknown   
2  000229ec398224ef6ca0657da4fc703e            0.0         Unknown   
3  00024acbcdf0a6daa1e931b038114c75            0.0         Unknown   
4  00042b26cf59d7ce69dfabb4e55b4fd9            0.0         Unknown   

   delivery_key  
0           455  
1           455  
2           455  
3           455  
4           455  

Missing delivery keys: 0


Next: Risk Key

Ab dim_risk mein bhi Unknown record add karenge, bilkul marketing aur delivery ki tarah.

In [67]:
# ==========================================
# ADD UNKNOWN RISK RECORD
# ==========================================

with dw_conn.cursor() as cursor:
    cursor.execute("""
        INSERT INTO dim_risk
        (
            fraud_risk_score,
            return_flag,
            customer_lifetime_value,
            risk_category
        )
        VALUES (0, 0, 0, 'Unknown')
    """)

dw_conn.commit()

print("Unknown risk record added successfully!")

Unknown risk record added successfully!


In [68]:
# ==========================================
# RETRIEVE RISK KEYS
# ==========================================

risk_keys = pd.read_sql(
    """
    SELECT
        risk_key,
        fraud_risk_score,
        return_flag,
        customer_lifetime_value,
        risk_category
    FROM dim_risk
    """,
    dw_conn
)

print("Risk dimension rows:", len(risk_keys))
print(risk_keys.tail())

Risk dimension rows: 1001
      risk_key  fraud_risk_score  return_flag  customer_lifetime_value  \
996        997              0.31            0                   1494.0   
997        998              0.60            0                   6062.0   
998        999              0.87            0                   4253.0   
999       1000              0.07            0                   4445.0   
1000      1001              0.00            0                      0.0   

     risk_category  
996         Medium  
997         Medium  
998           High  
999            Low  
1000       Unknown  


In [69]:
# ==========================================
# REMAP RISK KEY
# ==========================================

fact_sales_data = fact_sales_data.drop(
    columns=["risk_key"],
    errors="ignore"
)

fact_sales_data = fact_sales_data.merge(
    risk_keys,
    on=[
        "fraud_risk_score",
        "return_flag",
        "customer_lifetime_value",
        "risk_category"
    ],
    how="left"
)

print("Risk key mapped successfully!")

print(
    fact_sales_data[
        [
            "order_id",
            "fraud_risk_score",
            "risk_category",
            "risk_key"
        ]
    ].head()
)

print(
    "\nMissing risk keys:",
    fact_sales_data["risk_key"].isna().sum()
)

Risk key mapped successfully!
                           order_id  fraud_risk_score risk_category  risk_key
0  00010242fe8c5a6d1ba2dd792cb16214               0.0       Unknown    1001.0
1  00018f77f2f0320c557190d7a144bdd3               0.0       Unknown    1001.0
2  000229ec398224ef6ca0657da4fc703e               0.0       Unknown    1001.0
3  00024acbcdf0a6daa1e931b038114c75               0.0       Unknown    1001.0
4  00042b26cf59d7ce69dfabb4e55b4fd9               0.0       Unknown    1001.0

Missing risk keys: 143


In [70]:
# ==========================================
# CHECK UNMATCHED RISK ROWS
# ==========================================

unmatched_risk = fact_sales_data[
    fact_sales_data["risk_key"].isna()
][
    [
        "fraud_risk_score",
        "return_flag",
        "customer_lifetime_value",
        "risk_category"
    ]
]

print("Unmatched risk rows:", len(unmatched_risk))

print("\nFirst 20 unmatched rows:")
print(unmatched_risk.head(20))

Unmatched risk rows: 143

First 20 unmatched rows:
       fraud_risk_score  return_flag  customer_lifetime_value risk_category
656                0.60          1.0                   2047.0        Medium
657                0.60          1.0                   2047.0        Medium
2998               0.70          0.0                   8030.0        Medium
6900               0.95          1.0                   2203.0          High
9137               0.95          1.0                   8653.0          High
9596               0.47          1.0                   4407.0        Medium
11192              0.30          1.0                   5438.0           Low
11432              0.69          0.0                   1529.0        Medium
11800              0.47          0.0                   2095.0        Medium
11898              0.30          0.0                   2589.0           Low
12072              0.35          1.0                   8481.0        Medium
12652              0.57          0.0 

In [71]:
# ==========================================
# CHECK ONE UNMATCHED RISK COMBINATION
# ==========================================

check_risk = risk_keys[
    (risk_keys["fraud_risk_score"] == 0.60) &
    (risk_keys["return_flag"] == 1) &
    (risk_keys["customer_lifetime_value"] == 2047) &
    (risk_keys["risk_category"] == "Medium")
]

print(check_risk)

     risk_key  fraud_risk_score  return_flag  customer_lifetime_value  \
130       131               0.6            1                   2047.0   

    risk_category  
130        Medium  


In [72]:
# ==========================================
# CHECK FLOAT PRECISION
# ==========================================

fact_value = fact_sales_data.loc[656, "fraud_risk_score"]
dim_value = risk_keys.loc[
    (risk_keys["risk_key"] == 131),
    "fraud_risk_score"
].iloc[0]

print("Fact value:", repr(fact_value))
print("Dimension value:", repr(dim_value))
print("Are they exactly equal?:", fact_value == dim_value)
print("Difference:", fact_value - dim_value)

Fact value: np.float64(0.6000000000000001)
Dimension value: np.float64(0.6)
Are they exactly equal?: False
Difference: 1.1102230246251565e-16


In [73]:
# ==========================================
# FIX FLOATING-POINT PRECISION
# ==========================================

fact_sales_data["fraud_risk_score"] = (
    fact_sales_data["fraud_risk_score"].round(2)
)

risk_keys["fraud_risk_score"] = (
    risk_keys["fraud_risk_score"].round(2)
)

print("Fraud risk scores rounded successfully!")

print(
    "Example fact value:",
    repr(fact_sales_data.loc[656, "fraud_risk_score"])
)

print(
    "Example dimension value:",
    repr(
        risk_keys.loc[
            risk_keys["risk_key"] == 131,
            "fraud_risk_score"
        ].iloc[0]
    )
)

Fraud risk scores rounded successfully!
Example fact value: np.float64(0.6)
Example dimension value: np.float64(0.6)


In [74]:
# ==========================================
# REMAP RISK KEY AFTER PRECISION FIX
# ==========================================

fact_sales_data = fact_sales_data.drop(
    columns=["risk_key"],
    errors="ignore"
)

fact_sales_data = fact_sales_data.merge(
    risk_keys,
    on=[
        "fraud_risk_score",
        "return_flag",
        "customer_lifetime_value",
        "risk_category"
    ],
    how="left"
)

print("Risk key remapped successfully!")

print(
    fact_sales_data[
        [
            "order_id",
            "fraud_risk_score",
            "risk_category",
            "risk_key"
        ]
    ].head()
)

print(
    "\nMissing risk keys:",
    fact_sales_data["risk_key"].isna().sum()
)

Risk key remapped successfully!
                           order_id  fraud_risk_score risk_category  risk_key
0  00010242fe8c5a6d1ba2dd792cb16214               0.0       Unknown      1001
1  00018f77f2f0320c557190d7a144bdd3               0.0       Unknown      1001
2  000229ec398224ef6ca0657da4fc703e               0.0       Unknown      1001
3  00024acbcdf0a6daa1e931b038114c75               0.0       Unknown      1001
4  00042b26cf59d7ce69dfabb4e55b4fd9               0.0       Unknown      1001

Missing risk keys: 0


________________
Ab next step hai final fact table prepare karna.

- Step 1 — Final Fact Columns

In [75]:
# ==========================================
# PREPARE FINAL FACT TABLE
# ==========================================

final_fact = fact_sales_data[
    [
        "order_id",
        "customer_key",
        "product_key",
        "seller_key",
        "date_key",
        "payment_key",
        "marketing_key",
        "delivery_key",
        "risk_key",
        "price",
        "freight_value",
        "quantity",
        "total_amount"
    ]
].copy()

print("Final fact table prepared!")

print("\nShape:", final_fact.shape)

print("\nColumns:")
print(final_fact.columns.tolist())

print("\nFirst 5 rows:")
print(final_fact.head())

Final fact table prepared!

Shape: (112650, 13)

Columns:
['order_id', 'customer_key', 'product_key', 'seller_key', 'date_key', 'payment_key', 'marketing_key', 'delivery_key', 'risk_key', 'price', 'freight_value', 'quantity', 'total_amount']

First 5 rows:
                           order_id  customer_key  product_key  seller_key  \
0  00010242fe8c5a6d1ba2dd792cb16214         65558        25866         514   
1  00018f77f2f0320c557190d7a144bdd3         34266        27231         472   
2  000229ec398224ef6ca0657da4fc703e         34956        22625        1825   
3  00024acbcdf0a6daa1e931b038114c75         51764        15404        2024   
4  00042b26cf59d7ce69dfabb4e55b4fd9          7603         8863        1598   

   date_key  payment_key  marketing_key  delivery_key  risk_key   price  \
0  20170913           29           1001           455      1001   58.90   
1  20170426           30           1001           455      1001  239.90   
2  20180114           31           1001          

Ab database mein load karne se pehle ek important validation karte hain: NULL/missing values aur data types check karenge.

In [76]:
# ==========================================
# FINAL FACT TABLE VALIDATION
# ==========================================

print("Missing values in final_fact:")
print(final_fact.isna().sum())

print("\nData types:")
print(final_fact.dtypes)

print("\nTotal rows:", len(final_fact))

Missing values in final_fact:
order_id         0
customer_key     0
product_key      0
seller_key       0
date_key         0
payment_key      0
marketing_key    0
delivery_key     0
risk_key         0
price            0
freight_value    0
quantity         0
total_amount     0
dtype: int64

Data types:
order_id          object
customer_key       int64
product_key        int64
seller_key         int64
date_key           int64
payment_key        int64
marketing_key      int64
delivery_key       int64
risk_key           int64
price            float64
freight_value    float64
quantity           int64
total_amount     float64
dtype: object

Total rows: 112650


In [77]:
# ==========================================
# CHECK EXISTING FACT TABLE
# ==========================================

fact_count = pd.read_sql(
    """
    SELECT COUNT(*) AS total_rows
    FROM fact_sales
    """,
    dw_conn
)

print(fact_count)

   total_rows
0           0


- Step — Load Fact Table

In [78]:
# ==========================================
# LOAD FINAL FACT TABLE
# ==========================================

from psycopg2.extras import execute_values

fact_rows = list(
    final_fact[
        [
            "order_id",
            "customer_key",
            "product_key",
            "seller_key",
            "date_key",
            "payment_key",
            "marketing_key",
            "delivery_key",
            "risk_key",
            "price",
            "freight_value",
            "quantity",
            "total_amount"
        ]
    ].itertuples(index=False, name=None)
)

fact_insert_query = """
    INSERT INTO fact_sales
    (
        order_id,
        customer_key,
        product_key,
        seller_key,
        date_key,
        payment_key,
        marketing_key,
        delivery_key,
        risk_key,
        price,
        freight_value,
        quantity,
        total_amount
    )
    VALUES %s
"""

with dw_conn.cursor() as cursor:
    execute_values(
        cursor,
        fact_insert_query,
        fact_rows,
        page_size=5000
    )

dw_conn.commit()

print("Fact table loaded successfully!")
print("Rows inserted:", len(fact_rows))

Fact table loaded successfully!
Rows inserted: 112650


- ek final database validation karte hain — confirm karenge ke PostgreSQL mein exactly 112,650 rows hain aur koi duplicate/missing foreign key issue nahi.

In [79]:
# ==========================================
# FINAL FACT TABLE VALIDATION
# ==========================================

fact_check = pd.read_sql(
    """
    SELECT
        COUNT(*) AS total_rows,
        COUNT(DISTINCT order_id) AS unique_orders,
        SUM(quantity) AS total_quantity,
        SUM(total_amount) AS total_sales
    FROM fact_sales
    """,
    dw_conn
)

print(fact_check)

   total_rows  unique_orders  total_quantity  total_sales
0      112650          98666          112650  15843553.24


#### Current warehouse result:

- Fact rows: 112,650
- Unique orders: 98,666
- Total quantity: 112,650
- Total sales: 15,843,553.24
- Multiple rows per order are expected because our grain is one product item per order.

Ab last ETL validation karte hain: foreign-key integrity. Isse confirm hoga ke fact table ka koi surrogate key apni dimension mein missing reference nahi kar raha.

- Step — Foreign Key Integrity Check

In [80]:
# ==========================================
# FOREIGN KEY INTEGRITY CHECK
# ==========================================

fk_check = pd.read_sql(
    """
    SELECT
        COUNT(*) AS total_fact_rows,

        COUNT(*) FILTER (
            WHERE c.customer_key IS NULL
        ) AS missing_customer,

        COUNT(*) FILTER (
            WHERE p.product_key IS NULL
        ) AS missing_product,

        COUNT(*) FILTER (
            WHERE s.seller_key IS NULL
        ) AS missing_seller,

        COUNT(*) FILTER (
            WHERE d.date_key IS NULL
        ) AS missing_date,

        COUNT(*) FILTER (
            WHERE pay.payment_key IS NULL
        ) AS missing_payment,

        COUNT(*) FILTER (
            WHERE m.marketing_key IS NULL
        ) AS missing_marketing,

        COUNT(*) FILTER (
            WHERE del.delivery_key IS NULL
        ) AS missing_delivery,

        COUNT(*) FILTER (
            WHERE r.risk_key IS NULL
        ) AS missing_risk

    FROM fact_sales f

    LEFT JOIN dim_customer c
        ON f.customer_key = c.customer_key

    LEFT JOIN dim_product p
        ON f.product_key = p.product_key

    LEFT JOIN dim_seller s
        ON f.seller_key = s.seller_key

    LEFT JOIN dim_date d
        ON f.date_key = d.date_key

    LEFT JOIN dim_payment pay
        ON f.payment_key = pay.payment_key

    LEFT JOIN dim_marketing m
        ON f.marketing_key = m.marketing_key

    LEFT JOIN dim_delivery del
        ON f.delivery_key = del.delivery_key

    LEFT JOIN dim_risk r
        ON f.risk_key = r.risk_key
    """
    ,
    dw_conn
)

print(fk_check)

   total_fact_rows  missing_customer  missing_product  missing_seller  \
0           112650                 0                0               0   

   missing_date  missing_payment  missing_marketing  missing_delivery  \
0             0                0                  0                 0   

   missing_risk  
0             0  


### 🎉 ETL + Data Warehouse loading successfully complete.

Check the final integrity of warehouse 

| Check                |      Result |
| -------------------- | ----------: |
| Fact rows            | **112,650** |
| Missing Customer FK  |       **0** |
| Missing Product FK   |       **0** |
| Missing Seller FK    |       **0** |
| Missing Date FK      |       **0** |
| Missing Payment FK   |       **0** |
| Missing Marketing FK |       **0** |
| Missing Delivery FK  |       **0** |
| Missing Risk FK      |       **0** |
